### PR TITLE
[WIP] Transparent Torification support

### PR DESCRIPTION
**Pull request by Jookia**
I'm not familiar with the codebase or Android development in general, but this seems to work. I copypasted the VPN implementation to add a new checkbox that pushes traffic through Orbot. Works on my phone.

Is this a viable feature to be added? If so I'll clean it up.

### CODE REVIEW

**Comment by ukanth** [comment]
Thanks for the PR. I can see only UI changes. Where are the iptables rules for Tor/Orbot ? Or it just a UI for Orbot VPN ?

**Comment by Jookia** [comment]
Thanks for the reply!
It adds  the rules to the NAT table, so on my phone it outputs like this:

```
i9100:/ # iptables -t nat -L
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         
oem_nat_pre  all  --  anywhere             anywhere            

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         
afwall-tor  all  --  anywhere             anywhere            
afwall-tor  all  --  anywhere             anywhere            
afwall-tor  all  --  anywhere             anywhere            
afwall-tor  all  --  anywhere             anywhere            
afwall-tor  all  --  anywhere             anywhere            
afwall-tor  all  --  anywhere             anywhere            
afwall-tor  all  --  anywhere             anywhere            
afwall-tor  all  --  anywhere             anywhere            
afwall-tor  all  --  anywhere             anywhere            
afwall-tor  all  --  anywhere             anywhere            
afwall-tor  all  --  anywhere             anywhere            
afwall-tor  all  --  anywhere             anywhere            
afwall-tor  all  --  anywhere             anywhere            

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
natctrl_nat_POSTROUTING  all  --  anywhere             anywhere            

Chain afwall-tor (13 references)
target     prot opt source               destination         
REDIRECT   tcp  -- !localhost           !localhost            owner UID match u0_a9 tcp flags:FIN,SYN,RST,ACK/SYN redir ports 9040
REDIRECT   udp  --  anywhere             anywhere             owner UID match u0_a9 udp dpt:domain redir ports 5400
REDIRECT   tcp  -- !localhost           !localhost            owner UID match u0_a50 tcp flags:FIN,SYN,RST,ACK/SYN redir ports 9040
REDIRECT   udp  --  anywhere             anywhere             owner UID match u0_a50 udp dpt:domain redir ports 5400
REDIRECT   tcp  -- !localhost           !localhost            owner UID match u0_a53 tcp flags:FIN,SYN,RST,ACK/SYN redir ports 9040
REDIRECT   udp  --  anywhere             anywhere             owner UID match u0_a53 udp dpt:domain redir ports 5400
REDIRECT   tcp  -- !localhost           !localhost            owner UID match u0_a70 tcp flags:FIN,SYN,RST,ACK/SYN redir ports 9040
REDIRECT   udp  --  anywhere             anywhere             owner UID match u0_a70 udp dpt:domain redir ports 5400
REDIRECT   tcp  -- !localhost           !localhost            owner UID match u0_a72 tcp flags:FIN,SYN,RST,ACK/SYN redir ports 9040
REDIRECT   udp  --  anywhere             anywhere             owner UID match u0_a72 udp dpt:domain redir ports 5400
REDIRECT   tcp  -- !localhost           !localhost            owner UID match u0_a75 tcp flags:FIN,SYN,RST,ACK/SYN redir ports 9040
REDIRECT   udp  --  anywhere             anywhere             owner UID match u0_a75 udp dpt:domain redir ports 5400
REDIRECT   tcp  -- !localhost           !localhost            owner UID match u0_a88 tcp flags:FIN,SYN,RST,ACK/SYN redir ports 9040
REDIRECT   udp  --  anywhere             anywhere             owner UID match u0_a88 udp dpt:domain redir ports 5400
REDIRECT   tcp  -- !localhost           !localhost            owner UID match u0_a97 tcp flags:FIN,SYN,RST,ACK/SYN redir ports 9040
REDIRECT   udp  --  anywhere             anywhere             owner UID match u0_a97 udp dpt:domain redir ports 5400
REDIRECT   tcp  -- !localhost           !localhost            owner UID match u0_a98 tcp flags:FIN,SYN,RST,ACK/SYN redir ports 9040
REDIRECT   udp  --  anywhere             anywhere             owner UID match u0_a98 udp dpt:domain redir ports 5400
REDIRECT   tcp  -- !localhost           !localhost            owner UID match u0_a103 tcp flags:FIN,SYN,RST,ACK/SYN redir ports 9040
REDIRECT   udp  --  anywhere             anywhere             owner UID match u0_a103 udp dpt:domain redir ports 5400

Chain natctrl_nat_POSTROUTING (1 references)
target     prot opt source               destination         

Chain oem_nat_pre (1 references)
target     prot opt source               destination
```
See https://github.com/Jookia/afwall/blob/9b070d9907177980ce472cb474a36f09e38244c6/aFWall/src/main/java/dev/ukanth/ufirewall/Api.java#798

I'm not sure if this would be better to use a jump to a redirect instead of doing it each rule, but I'm not sure how to do that.

After using this patch for a while I haven't found too many bugs- The only one I've noticed is that removing firewall rules doesn't remove the nat table so you can't un-Tor applications.

It would be good to be able to select the proxy + port in settings too.

**Comment by ukanth** [comment]
It should be part of afwall OUTPUT chain and then filter it based on connection type ( look at afwall-vpn and other chains) instead of directly using output chain.

Also new Orbot has VPN connection. How does this takes care that ?

**Comment by Jookia** [comment]
Does afwall have a NAT output chain? This is all done using the NAT table, not the standard table.

This doesn't do anything to the VPN connection, I haven't used the VPN connection.

**Comment by ukanth** [comment]
No, it doesn't have NAT. Orbot has VPN way as well.

**Comment by Jookia** [comment]
Ah. Well this using iptables' NAT table to push applications through Tor. Orbot does have a VPN mode, but I don't know how to use it. I'm a bit less interested in using it since I'd like to use OpenVPN eventually and it would conflict I think.

**Comment by ukanth** [comment]
I understand that you would like to use OpenVPN,My only concern is integrating in AFWall+, people might  get confuse when using orbot with and without vpn along with vpn option in afwall. I need to think how does this goes along with other use cases and features

**Comment by Jookia** [comment]
Ah okay. I think what Orbot does is run all your traffic through a tap node and redirect certain traffic to Tor, not iptables needed. What this patch does is more like Orwall where it uses iptables to do the redirection.

The use case for this patch is when you want to have the benefits of Orwall without losing the ability to filter traffic using AFWall+. I'm not sure how big a use case this is.
The alternate to this would be to have Orwall and AFWall+ somehow interact and not stomp over their firewall rules.

Another route for this would be making it so this could be a custom script?  I originally had it as a custom script for this but it was annoying because I couldn't select applications in AFWall easily. Maybe the custom rules feature could be expanded to output a list of selected UIDs? This would be discussion for #316 

It might also be relevant that Orwall is no longer maintained so I don't think there's a chance of adding compatibility on that end.

Thanks for listening and discussing this. :)

**Comment by ukanth** [comment]
Does your pull request have all required files ?

**Comment by Jookia** [comment]
Yes, I don't have an icon for Orbot yet so I'm using the VPN icon.
Are you having trouble compiling?

**Comment by ukanth** [comment]
No, I will go through the changes and check it. Just wanted to make sure because you have mentioned as WIP. Is there anything pending from your end on this request ? If any please push those changes as well.

**Comment by Jookia** [comment]
Ah, yes. This is the status is this:

- [x] Get a prototype done that works on my phone
- [x] Create a PR to see if its a good approach
- [x] Test it day-to-day on my phone for a while
- [ ] Make discussed changes:
  - [ ] Use a single chain that applications jump to in the table
  - [ ] Flush the NAT table on rule clear
- [ ] Clean up code style
- [ ] Add Orbot icon
- [ ] Allow choosing transproxy address and port

If this seems like a good plan I'll get started.
The last two will probably be the hardest since I have no GUI experience.

I haven't done further work on it yet because I don't want to find out I'm doing things completely wrong and have to rewrite a different way, so if you could check the code that would be really helpful.

File: aFWall/build.gradle [hunk 1]
```diff
@@ -31,7 +31,7 @@ android {
 
     lintOptions {
         disable 'MissingTranslation'
-        abortOnError true
+        abortOnError false
```
Comment (ukanth) [inline aFWall/build.gradle]: Please revert this change.

File: aFWall/src/main/java/dev/ukanth/ufirewall/Api.java [hunk 2]
```diff
@@ -781,6 +795,18 @@ private static boolean applyIptablesRulesImpl(final Context ctx, RuleDataSet rul
...
+            cmds.add("-t nat -F " + AFWALL_CHAIN_NAME + "-tor");
+            cmds.add("-t nat -A OUTPUT -j " + AFWALL_CHAIN_NAME + "-tor");
+            for (Integer uid : ruleDataSet.torList) {
+                if (uid != null && uid >= 0) {
+                    cmds.add("-t nat -A " + AFWALL_CHAIN_NAME + "-tor -p tcp ! -d 127.0.0.1 ! -s 127.0.0.1 -m owner --uid-owner " + uid + " -m tcp --syn -j REDIRECT --to-ports 9040");
+                    cmds.add("-t nat -A " + AFWALL_CHAIN_NAME + "-tor -p udp -m owner --uid-owner " + uid + " -m udp --dport 53 -j REDIRECT --to-ports 5400");
+                }
+            }
+            }
+
```
Comment (ukanth) [inline aFWall/src/main/java/dev/ukanth/ufirewall/Api.java:806]: You should use addRulesForUidlist which handles uid for each app and whitelist/blocklist.

**Review by ukanth — COMMENTED** [review]
Check my review comments. For Icon Use https://materialdesignicons.com/ (with tor search - xml based)
Also it's good to externalize/preference for ports.

**Comment by tlaurion** [comment]
This is more then needed since orwall is not [working](https://github.com/EthACKdotOrg/orWall/issues/121) anymore and is unmaintained

**Comment by Jookia** [comment]
Thanks for the feedback @ukanth , I've been meaning to rework the patch but having something hit my inbox was a good motivator. ;)

I worked on a patch today which I should have ready tomorrow (it's midnight where I am  at the moment.) It should address most the issues (including having an icon from the material site, thanks for the link! Is there a place I need to credit it in the source code?), though it's a little more complex as there's currently a leak with this patch: Applications are allowed to output any packets, but only a subset will go through Tor- TCP and port 53 UDP. My reworked patch will fix this by adding an extra filter step for traffic marked for Torified applications.

On another note, you wrote "You should use addRulesForUidlist which handles uid for each app and whitelist/blocklist". I spent a bit trying this, but the problem I've hit is that there's a lot of extra code I don't understand in it relating to DNS proxies and tethering. I don't want to write code that lets a subset of traffic through only for more to be let through. I'd also have to refactor the function a little to pass in the argument of a destination jump- All UIDs marked as being Torified need to jump to a 'tor-filter' chain which will mark some traffic and reject the rest to stop leaks.

At least, that's the only viable way that would seem to work. I haven't written it yet. ;)

**Comment by ukanth** [comment]
@Jookia , That's ok. Atleast make sure how does it handle two modes (whitelist/blocklist).  addRulesForUidlist applies the rules based on the mode.

**Comment by Jookia** [comment]
Ok, so I spent today rewriting the rules again and testing. I'm pretty confident that it works, so I've pushed the new version. Most of the details are in the commit message, so as a quick summary of differences:

- There's an icon now for the Tor checkboxes
- The checkbox column lets you put applications in 'Tor mode'
- Tor mode transparently proxies traffic
- Tor mode allows access to Orbot's proxies directly
- Tor mode will reject inbound, IPv6 and non-Torified traffic (non-DNS UDP)

AFWall+ changes:

- Added NAT chains (IPv6 only)
- Added a chain for INPUT
- Marked a bug in the source code (search the commit for BUG: )

I haven't been able to lint it on my machine because whatever SDK I'm using is the wrong API so it complains about those incompatibilities.

Most the work from here is probably GUI-related. It would be good to somehow indicate that Tor mode isn't about whitelisting/blacklisting traffic but imposing a 'mode' of the traffic. There also needs to be warning saying 'Applications restricted to Tor mode may still use other applications to connect directly to the Internet. An example of this is an application in Tor mode that uses Android's download manager  (not in Tor mode) to download files directly over the Internet.

There also needs to be a way to set the Orbot ports for its utilities in the GUI. I don't know how to set this up since most my Android coding is copy pasting.

Another issue is that this balloons the firewall rules (at least if you have inbound connections and IPv6 enabled), so it takes a little bit to process. On my phone I count 318 rules. I think most of them are uid checks. Maybe it'd be a good idea to use the mark system and bitmasks to fix this.

That's about it. If anyone wants a copy of the debug APK (or knows where to host it), I'm happy to provide it FOR TESTING PURPOSES- just message me. My email is in the git commit which you should read to understand what Tor mode does and doesn't do.

**Comment by tlaurion** [comment]
> If  anyone wants a copy of the debug APK (or knows where to host it), I'm  happy to provide it FOR TESTING PURPOSES- just message me. My email is  in the git commit which you should read to understand what Tor mode does  and doesn't do.

@Jookia : Please send the apk on [framadrop](https://framadrop.org/) or other online services and drop the link here with SHA512 sums. I would definitely give it a shot.

I'm trying to adapt [this](https://github.com/mikeperry-tor/mission-improbable/ ) with [that](https://blog.forkwhiletrue.me/posts/an-almost-fully-libre-galaxy-s3/) on Galaxy S3 i9300. Mission Improbable relies on [unmaintained orwall](https://github.com/EthACKdotOrg/orWall/issues/124) and even [using previous 1.1 version](https://github.com/EthACKdotOrg/orWall/issues/121) to force torification of defined traffic network communication from boot. [Afwall doesn't currently support VOIP passthrough out of the box without custom scripts ](https://github.com/ukanth/afwall/wiki/CustomScripts) while newer versions of Android breaks everything by creating DNS leaks which AFwall prevents.

@Jookia @mikeperry-tor @cjeanneret and @hgourvest: Should Orwall be maintained instead? 
@Jookia: Should Orwall features be merged into Afwall? Should efforts be merged?

**Comment by Jookia** [comment]
On Wed, Sep 12, 2018 at 08:03:40AM -0700, tlaurion wrote:
> > If  anyone wants a copy of the debug APK (or knows where to host it), I'm  happy to provide it FOR TESTING PURPOSES- just message me. My email is  in the git commit which you should read to understand what Tor mode does  and doesn't do.
> 
> @Jookia : Please send the apk on framadrop or other online services and drop the link here with SHA512 sums. I would definitely give it a shot.

Oh, framadrop's a thing? Here you go:

https://framadrop.org/r/9sYHbmJOk4#nb2OqeaYyaH2hKuquT1qSw0BjSVidua810ZB3BPkUVk=

47f76b67fca9a69ecb2e0b3834d630cd7463b65fcbcd6c52da6dc05e0bf04021070d26a59edf670ea1b58e43c860dd08d3d408272b446c3a461503ee3abf4071

> I'm trying to adapt [this](https://github.com/mikeperry-tor/mission-improbable/ ) with [that](https://blog.forkwhiletrue.me/posts/an-almost-fully-libre-galaxy-s3/) on Galaxy S3 i9300, which relies on [unmaintained orwall](https://github.com/EthACKdotOrg/orWall/issues/124) to force torification of defined traffic and blocking of everything else from boot. Afwall doesn't support VOIP passthrough out of the box and newer versions of Android breaks everything by creating DNS leaks, which AFwall prevents.

This AFwall patch works a bit like Orwall: You block/unblock traffic as
usual, then tell the unblocked traffic to get filtered through Tor.

So for an orWall experience you'd use a whitelist where you list a set
of applications to be able to access Wi-Fi, Cellular, and push it
through Tor. For apps that support Tor like F-Droid you just set it to
be pushed through Tor and it will be let through since it talks to
Orwall.

Make sure to let Orbot access the Internet without Tor.

> @Jookia @mikeperry-tor @cjeanneret and @hgourvest: Should Orwall be maintained instead? 

If possible, yes. It has years of testing and a simpler interface. I
don't think at any point it gives users the easy option to bypass Tor,
unlike a firewall that intends to let you do that. Using AFWall with
this patch to give a safe interface to avoid leaking traffic would not
be the greatest idea. To use a Linux analogy, 

I think the closest relative of Orwall would be Orbot's VPN mode
currently.

> @Jookia: Should Orwall features be merged into Afwall? Should efforts be merged?

I think Orwall and AFwall have two different audiences and goals. So I
don't think its possible to make a single application that does both.

I imagine there's some kind of functionality that would be nice to tease
out and share to allow making different firewall apps but avoiding
firewall busywork.

**Comment by tlaurion** [comment]
Works great. Testing it.
@Jookia: for some reason, framadrop doesn't work well with any of the mobile browser tested and I had to download on computer then push app on cell phone. If we want this to be tested, it would be nice if it was in the releases to be easily downloadable by people.

Will test leaks on wifi network and report back.

**Comment by ukanth** [comment]
@Jookia , Kindly resolve merge conflicts and resubmit. I will test it and merge the changes this weekend and early next week.

**Comment by Jookia** [comment]
I've rebased and done basic testing. Everything seems to be working fine.

**Comment by ukanth** [comment]
Looks good to me. Tested and works fine. If Orbot stopped and app is still routed traffic through tor, then it will be unable to connect to internet. This can be solved using profiles ( different profile just for tor). 

Great job @Jookia.

**Comment by Jookia** [comment]
Thanks :) I can't wait for the next release.

**Comment by ijesaboy** [comment]
> Looks good to me. Tested and works fine. If Orbot stopped and app is still routed traffic through tor, then it will be unable to connect to internet. This can be solved using profiles ( different profile just for tor).
> 
> Great job @Jookia.

@Jookia So for a given app:

1. When the wifi is enabled for the app and tor-control disabled, i get internet access.
2. When tor-control enabled and Orbot is running, i get internet access.
3. However, when I have wifi and tor-control enabled and Orbot is stopped (in this case it had not been running since reboot) I can't get internet access. 

My expectation upon opening the app in case#3 is that it would connect to internet via wifi but it doesn't. Is this a limitation of the tor-control feature that ukanth described above or is there something else afoot in my instance. 

Note: My afwall log view shows no blocked wifi connections for the app; please find rules & log linked below:
https://gist.github.com/ijesaboy/4fa220600e059a2f3f4fb85fb8f6e2fe

Can you recommend a custom script of fix that would connect in the 3rd case above? 
Many thanks for taking a look.

**Comment by Jookia** [comment]
It's not that there's any blocked connections, it's just that traffic is being routed to Orbot which isn't running.
Overall this is deliberate so that you don't accidentally leak sensitive data over Wi-Fi or mobile data if Orbot stops.

**Comment by ijesaboy** [comment]
The desire to avoid leaks makes sense. Many thanks for the quick followup :) 
Is it then fair to say that if planning to use tor-control, one should only enable tor-control option since the other data options would be mute or represent potential leaks?

If so, ukanth's comment makes more sense and I'll get to setting up an alternate profile for the 'orbot is stopped' case.

**Comment by Jookia** [comment]
> Is it then fair to say that if planning to use tor-control, one should only enable tor-control option since the other data options would be mute or represent potential leaks?

It depends if the traffic filtering is done before or after the redirection. It would either do nothing (if filtering traffic after redirecting through Tor) or stop traffic being redirected through Tor (if filtering is done before redirection). I don't remember the details.

**Comment by ijesaboy** [comment]
Got it. I can use iptables commands to check, yes? If the redirect happens further down the chain then it is applied later relative to the rules higher up the chain, yes?

Also, can't thank you enough for the follow up. I'm a newbie and your responses are super clarifying.

**Comment by ijesaboy** [comment]
@Jookia one last question for you (that you've probably already answered).
For afwall Core & System apps, should I explicitly NOT use tor-control since it might redirect the data in ways that could prevent connections from other enabled modes? or is it OK to enable it in several cases?

A few specific ones I was : 
- **linux kernel** [-11] :: Things seem to be work fine with tor-control + I was under the impression every mode should be enabled.  
- **vpn networking** [1016] or **tethering** [-12] :: maybe to enable vpn/tethering over tor as well as other modes? 
- **Android System** [1000] :: I was under the impression every mode but LAN should be enabled
- **PacProcessor** & **Intent Filter** :: a couple more I was under the impression every mode but LAN should be enabled

I'm trying to test these now but having difficulty getting tethering to work.

**Comment by Jookia** [comment]
Feel free to ask as many questions as you want :)

I'm not too sure about the risks of transparently proxying things through Tor. I do know that you want to avoid proxying TCP over Tor, since that will cause a TCP meltdown.
